src/irmin-pack/unix/ext.ml

```diff
@@ -128,100 +124,92 @@ module Maker (Config : Conf.S) = struct
+            in
+            raise exn
           in
-          let* contents =
-            Contents.CA.v ~readonly ~lru_size ~index ~indexing_strategy ~dict
-              ~io
-          in
-          let* node =
-            Node.CA.v ~readonly ~lru_size ~index ~indexing_strategy ~dict ~io
+          Lwt.try_bind (fun () -> f contents node commit) on_success on_fail
+
+        let v config =
+          let fm =
+            let readonly = Irmin_pack.Conf.readonly config in
+            (* TODO: Proper exceptions (instead of [Result.get_ok]) *)
+            if readonly then File_manager.open_ro config |> Result.get_ok
+            else
+              let fresh = Irmin_pack.Conf.fresh config in
+              let root = Irmin_pack.Conf.root config in
+              match (Io.classify_path root, fresh) with
+              | `No_such_file_or_directory, _ ->
+                  File_manager.create_rw ~overwrite:false config
+                  |> Result.get_ok
+              | `Directory, true ->
+                  File_manager.create_rw ~overwrite:true config |> Result.get_ok
+              | `Directory, false ->
+                  File_manager.open_rw config |> Result.get_ok
+              | `File, _ ->
+                  (* TODO: Proper exception *)
+                  assert false
           in
-          let* commit =
-            Commit.CA.v ~readonly ~lru_size ~index ~indexing_strategy ~dict ~io
+          let dict =
+            (* TODO: Capacity from config? *)
+            let capacity = 100_000 in
+            Dict.v ~capacity fm
           in
-
+          let* contents = Contents.CA.v ~config ~fm ~dict in
+          let* node = Node.CA.v ~config ~fm ~dict in
+          let* commit = Commit.CA.v ~config ~fm ~dict in
           let+ branch =
-            let path = Irmin_pack.Layout.branch ~root in
+            let root = Conf.root config in
+            let fresh = Conf.fresh config in
+            let readonly = Conf.readonly config in
+            let path = Irmin_pack.Layout.V3.branch ~root in
             Branch.v ~fresh ~readonly path
           in
-          (* Stores share instances in memory, one flush is enough. In case of a
-             system crash, the flush_callback might not make with the disk. In
-             this case, when the store is reopened, [integrity_check] needs to be
-             called to repair the store. *)
-          (f := fun () -> Contents.CA.flush ~index:false contents);
-          { contents; node; commit; branch; config; index; dict; io }
+          { config; contents; node; commit; branch; fm; dict }
 
         let close t =
-          Index.close t.index;
-          [%log.debug "[pack] close %s" (Io_legacy.name t.io)];
-          Io_legacy.close t.io;
-          Dict.close t.dict;
-          Contents.CA.close (contents_t t) >>= fun () ->
-          Node.CA.close (snd (node_t t)) >>= fun () ->
-          Commit.CA.close (snd (commit_t t)) >>= fun () -> Branch.close t.branch
+          let () =
```

Comment:
I've added the following comment in the interface of the close:
```
  val close : t -> (unit, [> Io.close_error | `Pending_flush | `Tmp ]) result
  (** Close all the files.

      This call fails if the append buffers are not in a flushed stated. This
      situation will most likely never occur because the append buffers will
      contain data only during the scope of a batch function. *)
```